internal/graph/check.go

```diff
@@ -805,64 +848,181 @@ func buildTupleKeyConditionFilter(ctx context.Context, reqCtx *structpb.Struct,
+//
+// type user
+// type org
+//
+//	relations
+//		define viewer: [user]
+//
+// type folder
+//
+//	relations
+//		define viewer: [user]
+//
+// type doc
+//
+//	relations
+//		define viewer: viewer from parent
+//		define parent: [folder, org]
+//
+// works as follows.
+// If the request is Check(user:maria, viewer, doc:1).
+// 1. We build a map with folder#viewer:[1...N], org#viewer:[1...M] that are parents of doc:1. We send those through a channel.
+// 2. The consumer of the channel finds all the folders (and orgs) by looking at tuples of the form folder:X#viewer@user:maria (and org:Y#viewer@user:maria).
+// 3. If there is one folder or org found in step (2) that appears in the map found in step (1), it returns allowed=true immediately.
+func (c *LocalChecker) checkMembership(ctx context.Context, req *ResolveCheckRequest, iter *storage.ConditionsFilteredTupleKeyIterator, usersetDetails usersetDetailsFunc) (*ResolveCheckResponse, error) {
+	ctx, span := tracer.Start(ctx, "checkMembership")
+	defer span.End()
 
+	// since this is an unbuffered channel, producer will be blocked until consumer catches up
+	// TODO: when implementing set math operators, change to buffered. consider using the number of sets as the concurrency limit
+	usersetsChan := make(chan usersetsChannelType)
+
+	cancellableCtx, cancelFunc := context.WithCancel(ctx)
+	// sending to channel in batches up to a pre-configured value to subsequently checkMembership for.
+	pool := concurrency.NewPool(cancellableCtx, 1)
+	defer func() {
+		cancelFunc()
+		// We need to wait always to avoid a goroutine leak.
+		_ = pool.Wait()
+	}()
+	pool.Go(func(ctx context.Context) error {
+		c.produceUsersets(ctx, usersetsChan, iter, usersetDetails)
+		return nil
+	})
+
+	var finalErr error
+
+ConsumerLoop:
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case newBatch, channelOpen := <-usersetsChan:
+			if !channelOpen {
+				break ConsumerLoop
+			}
+			if newBatch.err != nil {
+				// Irrecoverable error when fetching usersets, so we abort.
+				finalErr = newBatch.err
+				break ConsumerLoop
+			}
+			objectRel := newBatch.objectRelation
+			objectIDs := newBatch.objectIDs
+
+			resp, err := c.buildCheckAssociatedObjects(req, objectRel, objectIDs)(ctx)
+			if err != nil {
+				// We don't exit because we do a best effort to find the objectId that will give `allowed=true`.
+				// If that doesn't happen, we will return this error down below.
+				finalErr = err
+			} else if resp.Allowed {
+				resp.GetResolutionMetadata().DatastoreQueryCount++
+				return resp, nil
+			}
+		}
+	}
+
+	// context cancellation from upstream (e.g. client)
+	if ctx.Err() != nil {
+		finalErr = ctx.Err()
+	}
+
+	if finalErr != nil {
+		telemetry.TraceError(span, finalErr)
+		return nil, finalErr
+	}
+
+	return &ResolveCheckResponse{
+		Allowed: false,
+		ResolutionMetadata: &ResolveCheckResponseMetadata{
+			DatastoreQueryCount: req.GetRequestMetadata().DatastoreQueryCount,
```

Comment:
```suggestion
			DatastoreQueryCount: req.GetRequestMetadata().DatastoreQueryCount++,
```

The reason being, we still did queries, just doesn't mean that we found anything.

Reply:
If anything we should increase the counter per each `objectRel` that we observed or per each batch we sent over the channel.

Reply:
In hindsight, passing this counter around on every request and response was a very bad idea :(